oak_client/src/lib.rs

```diff
@@ -22,34 +22,92 @@ pub mod proto {
 pub mod transport;
 pub mod verifier;
 
-use anyhow::Context;
+use crate::transport::AsyncEvidenceProvider;
+use anyhow::{anyhow, Context};
+use oak_crypto::{
+    schema::{AeadEncryptedMessage, HpkeRequest, HpkeResponse},
+    SenderCryptoProvider,
+};
+use prost::Message;
 use std::vec::Vec;
 
+const EMPTY_ASSOCIATED_DATA: &[u8] = b"";
+
 /// Client for connecting to Oak.
 /// Represents a Relying Party from the RATS Architecture:
 /// <https://www.rfc-editor.org/rfc/rfc9334.html#name-relying-party>
 pub struct OakClient<T>
 where
-    T: micro_rpc::AsyncTransport<Error = anyhow::Error>,
+    T: micro_rpc::AsyncTransport<Error = anyhow::Error> + AsyncEvidenceProvider,
 {
     transport: T,
+    crypto_provider: SenderCryptoProvider,
 }
 
 impl<T> OakClient<T>
 where
-    T: micro_rpc::AsyncTransport<Error = anyhow::Error>,
+    T: micro_rpc::AsyncTransport<Error = anyhow::Error> + AsyncEvidenceProvider,
 {
-    pub async fn create(transport: T) -> anyhow::Result<Self> {
+    pub async fn create(mut transport: T) -> anyhow::Result<Self> {
         // TODO(#3442): Fetch enclave public key.
         // TODO(#3641): Implement client-side attestation verification.
-        Ok(Self { transport })
+        let evidence = transport
+            .get_evidence()
+            .await
+            .context("couldn't get attestation evidence")?;
+        let crypto_provider = SenderCryptoProvider::new(&evidence.encryption_public_key);
+        Ok(Self {
+            transport,
+            crypto_provider,
+        })
     }
 
     pub async fn invoke(&mut self, request_body: &[u8]) -> anyhow::Result<Vec<u8>> {
-        // TODO(#3767): Use hybrid encryption for messages.
-        self.transport
-            .invoke(request_body)
+        // Create encryptor.
+        let (serialized_encapsulated_public_key, request_encryptor) = self
+            .crypto_provider
+            .create_encryptor()
+            .context("couldn't create encryptor")?;
+
+        // Encrypt and serialize request.
+        let (request_ciphertext, response_decryptor) = request_encryptor
```

Comment:
Why is the decryptor returned by the encryptor? Would it not be possible to return both encryptor and decryptor at the same time from the crypto provider?